src/main/java/com/tterrag/k9/mappings/official/OfficialDatabase.java

```diff
@@ -0,0 +1,118 @@
+                    .filter(s -> !s.startsWith("#"))
+                    .map(s -> s.replace('.', '/'))
+                    .collect(Collectors.toList());
+            populateMappings(mappings, lines, "client.txt".equals(path.getFileName().toString()));
+        }
+        return mappings;
+    }
+
+    // Derived from SrgUtils in InternalUtils#loadProguard
+    private void populateMappings(Set<OfficialMapping> mappings, List<String> lines, boolean isClient) throws IOException {
+        McpMapping.Side side = isClient ? McpMapping.Side.CLIENT : McpMapping.Side.SERVER;
+        OfficialMapping clazz = null;
+        for (String line : lines) {
+            if (!line.startsWith("    ") && line.endsWith(":")) {
+                String[] mapped = line.substring(0, line.length() - 1).split(" -> ");
+                clazz = addMapping(mappings, new OfficialMapping(srgs, this, side, MappingType.CLASS, null, null, mapped[1], mapped[0], null));
+            } else if (line.contains("(") && line.contains(")")) {
+                if (clazz == null)
+                    throw new IOException("Class was null when parsing method");
+
+                line = line.trim();
+                if (line.indexOf(':') != -1) {
+                    int i = line.indexOf(':');
+                    int j = line.indexOf(':', i + 1);
+                    line = line.substring(j + 1);
+                }
+
+                String original = line.split(" -> ")[1];
+                int spaceIndex = line.indexOf(' ');
+                String returnType = toDesc(line.substring(0, spaceIndex));
+                String name = line.substring(spaceIndex + 1, line.indexOf('('));
+                String[] args = line.substring(line.indexOf('(') + 1, line.indexOf(')')).split(",");
+
+                StringBuilder desc = new StringBuilder("(");
+                for (String arg : args) {
+                    if (arg.isEmpty())
+                        break;
+                    desc.append(toDesc(arg));
+                }
+                desc.append(')').append(returnType);
+                addMapping(mappings, new OfficialMapping(srgs, this, side, MappingType.METHOD, clazz, desc.toString(), original, name, null));
+            } else {
+                if (clazz == null)
+                    throw new IOException("Class was null when parsing field");
+
+                String[] pts = line.trim().split(" ");
+                addMapping(mappings, new OfficialMapping(srgs, this, side, MappingType.FIELD, clazz, null, pts[3], pts[1], pts[0]));
+            }
+        }
+    }
+
+    private OfficialMapping addMapping(Set<OfficialMapping> mappings, OfficialMapping toAdd) {
+        OfficialMapping toReturn = toAdd;
+        if (!mappings.add(toAdd)) {
+            mappings.remove(toAdd);
+            toReturn = toAdd.cloneWithSide(McpMapping.Side.BOTH);
+            mappings.add(toReturn);
+        }
+        return toReturn;
+    }
+
+    // Copied from SrgUtils in InternalUtils#toDesc
```

Comment:
This seems like bad design. How should I properly move this to use the clojure types and signatures so that it's less hardcoded?